cpp/src/arrow/compute/exec/spilling_join.cc

```diff
@@ -0,0 +1,351 @@
+                partition_idx_,
+                static_cast<size_t>(batch_id));
+            return builder_->PushNextBatch(
+                thread_index,
+                static_cast<int64_t>(num_rows),
+                hashes);
+        }
+
+        Status SpillingHashJoin::BuildNextBloomFilter(size_t thread_index)
+        {
+            if(partition_idx_ >= SpillingAccumulationQueue::kNumPartitions)
+                return OnBloomFilterFinished(thread_index);
+
+            size_t num_rows = build_accumulator_.row_count(partition_idx_);
+            size_t num_batches = build_accumulator_.batch_count(partition_idx_);
+
+            // partition_idx_ is incremented in the callback for the taskgroup
+            bloom_filter_->partitions[partition_idx_] =
+                arrow::internal::make_unique<BlockedBloomFilter>();
+
+            RETURN_NOT_OK(builder_->Begin(
+                              num_threads_,
+                              ctx_->cpu_info()->hardware_flags(),
+                              ctx_->memory_pool(),
+                              num_rows,
+                              num_batches,
+                              bloom_filter_->partitions[partition_idx_].get()));
+                
+            return callbacks_.start_task_group(
+                task_group_bloom_[partition_idx_],
+                build_accumulator_.batch_count(partition_idx_));
+        }
+        
+        Status SpillingHashJoin::OnBloomFilterFinished(size_t thread_index)
+        {
+            RETURN_NOT_OK(callbacks_.bloom_filter_finished(thread_index));
+            bool expected = false;
+            if(!bloom_or_probe_finished_.compare_exchange_strong(expected, true))
+                return StartCollocatedJoins(thread_index);
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::OnBloomFiltersReceived(size_t thread_index)
+        {
+            bloom_ready_.store(true, std::memory_order_release);
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::OnProbeSideBatch(size_t thread_index, ExecBatch batch)
+        {
+            if(bloom_ready_.load())
+            {
+                RETURN_NOT_OK(callbacks_.apply_bloom_filter(
+                    thread_index,
+                    &batch));
+            }
+            return probe_accumulator_.InsertBatch(
+                              thread_index,
+                              std::move(batch));
+        }
+
+        Status SpillingHashJoin::OnProbeSideFinished(size_t thread_index)
+        {
+            bool expected = false;
+            if(!bloom_or_probe_finished_.compare_exchange_strong(expected, true))
+                return StartCollocatedJoins(thread_index);
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::StartCollocatedJoins(size_t thread_index)
+        {
+            partition_idx_ = SpillingAccumulationQueue::kNumPartitions;
```

Comment:
You can add a comment saying: we read partitions in reversed order of writing, so that we may benefit more from any IO caches that may be (OS, SSD).

Reply:
Added